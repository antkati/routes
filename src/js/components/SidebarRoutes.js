import React, { useMemo } from 'react';
import { DragDropContext, Droppable } from 'react-beautiful-dnd';
import '../styles/map.css';
import SidebarRoutesItem from './SidebarRoutesItem.js';


let SidebarRoutes = (props) => {
    const { routes, handleInput, handleSidebarItemDelete, handleSidebarItemDrag } = props;

    const column = routes.columns['column-1'];

    const draggableItem = column.pointsIds.map((id, index) => {
        const point = routes.points[id];
        return <SidebarRoutesItem id={point.id} content={point.content} index={index} key={point.id} />
    });

    const droppableList = useMemo(
        () => (
            <DragDropContext onDragEnd={handleSidebarItemDrag}>
                <Droppable droppableId={column.id}>
                    {provided => (
                        <ul
                            className='map-sidebar__list'
                            onClick={handleSidebarItemDelete}
                            ref={provided.innerRef}
                            {...provided.droppableProps}
                        >
                            {draggableItem}
                            {provided.placeholder}
                        </ul>
                    )}
                </Droppable>
            </DragDropContext>
        ),
        [draggableItem]
    )

    return (
        <aside className='map-sidebar'>
            <input className='map-sidebar__input' type='text' placeholder='Введите название точки' onKeyDown={handleInput} />
            {droppableList}
        </aside>
    )
}

SidebarRoutes = React.memo(SidebarRoutes);

export default SidebarRoutes;