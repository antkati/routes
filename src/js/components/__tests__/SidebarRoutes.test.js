import React from 'react';
import { mount } from 'enzyme';
import SidebarRoutes from '../SidebarRoutes.js';

describe('SidebarRoutes', () => {
    const routesWithPoints = {
        points: {
            'point-0': { id: 'point-0', coords: [56.02279830, 92.89742900], content: 'точка 1' },
            'point-1': { id: 'point-1', coords: [56.02279830, 92.89742900], content: 'точка 2' },
            'point-2': { id: 'point-2', coords: [56.02279830, 92.89742900], content: 'точка 3' },
            'point-3': { id: 'point-3', coords: [56.02279830, 92.89742900], content: 'точка 4' },
        },
        columns: {
            'column-1': {
                id: 'column-1',
                pointsIds: ['point-0', 'point-1', 'point-2', 'point-3'],
            }
        },
        columnOrder: ['column-1'],
    }

    const routesWithoutPoints = {
        points: {},
        columns: {
            'column-1': {
                id: 'column-1',
                pointsIds: [],
            }
        },
        columnOrder: ['column-1'],
    }    

    const handleInput = jest.fn();
    const handleSidebarItemDelete = jest.fn();
    const handleSidebarItemDrag = () => {};

    const propsWithPoints = {
        routes: routesWithPoints,
        handleInput,
        handleSidebarItemDelete,
        handleSidebarItemDrag,
    }

    const propsWithoutPoints = {
        routes: routesWithoutPoints,
        handleInput,
        handleSidebarItemDelete,
        handleSidebarItemDrag,
    }

    let renderedComponent;

    function mountComponentWithPoints() {
        renderedComponent = mount(<SidebarRoutes {...propsWithPoints} />);
    }

    test('SidebarRoutes component renders correctly', () => {
        mountComponentWithPoints();
        
        expect(renderedComponent).toMatchSnapshot();
    });


    test('number of lis is equal to number of points', () => {
        mountComponentWithPoints();
        const lis = renderedComponent.find('.map-sidebar__list__item');

        const column = routesWithPoints.columns['column-1'];
        const pointsLength = column.pointsIds.length;

        expect(lis.length).toBe(pointsLength);
    });
    

    test('lis do not exist without points', () => {
        const renderedComponent = mount(<SidebarRoutes {...propsWithoutPoints} />);

        const lis = renderedComponent.find('.map-sidebar__list__item');
        expect(lis.length).toBe(0);
    }); 
 

    describe('render lis in right order', () => {
        mountComponentWithPoints();

        const column = routesWithPoints.columns['column-1'];
        const points = column.pointsIds.map(point => {
            return routesWithPoints.points[point];
        });

        const lis = renderedComponent.find('.map-sidebar__list__item');

        points.forEach((point, index) => {
            describe(`li-${index} matches ${point.id}`, () => {
                const li = lis.at(index);

                test(`data-id attribute matches point-id`, () => {
                    expect(li.prop('data-id')).toBe(point.id);
                })

                test(`data-index attribute matches point-index`, () => {
                    expect(li.prop('data-index')).toBe(index);
                })  
                
                test(`innerHTML matches point-content`, () => {
                    const liContent = li.find('span').first().text();
                    expect(liContent).toBe(point.content);
                })                 
            })
        })
    });
    

    test('onKeyDown callback has been called', () => {
        mountComponentWithPoints();

        const input = renderedComponent.find('.map-sidebar__input');  
        input.simulate('keydown');
        expect(handleInput).toHaveBeenCalled();      
    });  
    

    test('onClick callback has been called', () => {
        mountComponentWithPoints();

        const closeItem = renderedComponent.find('.close-item').first();  
        closeItem.simulate('click');
        expect(handleSidebarItemDelete).toHaveBeenCalled();  
    }); 
});


