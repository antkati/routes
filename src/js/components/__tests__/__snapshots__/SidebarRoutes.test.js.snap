// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarRoutes SidebarRoutes component renders correctly 1`] = `
<Memo(SidebarRoutes)
  handleInput={[MockFunction]}
  handleSidebarItemDelete={[MockFunction]}
  handleSidebarItemDrag={[Function]}
  routes={
    Object {
      "columnOrder": Array [
        "column-1",
      ],
      "columns": Object {
        "column-1": Object {
          "id": "column-1",
          "pointsIds": Array [
            "point-0",
            "point-1",
            "point-2",
            "point-3",
          ],
        },
      },
      "points": Object {
        "point-0": Object {
          "content": "точка 1",
          "coords": Array [
            56.0227983,
            92.897429,
          ],
          "id": "point-0",
        },
        "point-1": Object {
          "content": "точка 2",
          "coords": Array [
            56.0227983,
            92.897429,
          ],
          "id": "point-1",
        },
        "point-2": Object {
          "content": "точка 3",
          "coords": Array [
            56.0227983,
            92.897429,
          ],
          "id": "point-2",
        },
        "point-3": Object {
          "content": "точка 4",
          "coords": Array [
            56.0227983,
            92.897429,
          ],
          "id": "point-3",
        },
      },
    }
  }
>
  <aside
    className="map-sidebar"
  >
    <input
      className="map-sidebar__input"
      onKeyDown={[MockFunction]}
      placeholder="Введите название точки"
      type="text"
    />
    <DragDropContext
      onDragEnd={[Function]}
    >
      <ErrorBoundary>
        <App
          onDragEnd={[Function]}
          setOnError={[Function]}
          uniqueId={1}
        >
          <Provider
            context={
              Object {
                "$$typeof": Symbol(react.context),
                "Consumer": Object {
                  "$$typeof": Symbol(react.context),
                  "_calculateChangedBits": null,
                  "_context": [Circular],
                },
                "Provider": Object {
                  "$$typeof": Symbol(react.provider),
                  "_context": [Circular],
                },
                "_calculateChangedBits": null,
                "_currentRenderer": Object {},
                "_currentRenderer2": null,
                "_currentValue": null,
                "_currentValue2": null,
                "_threadCount": 0,
              }
            }
            store={
              Object {
                "dispatch": [Function],
                "getState": [Function],
                "replaceReducer": [Function],
                "subscribe": [Function],
                Symbol(observable): [Function],
              }
            }
          >
            <ConnectFunction
              direction="vertical"
              droppableId="column-1"
              ignoreContainerClipping={false}
              isCombineEnabled={false}
              isDropDisabled={false}
              type="DEFAULT"
            >
              <Droppable
                direction="vertical"
                droppableId="column-1"
                ignoreContainerClipping={false}
                isCombineEnabled={false}
                isDropDisabled={false}
                placeholder={null}
                shouldAnimatePlaceholder={true}
                snapshot={
                  Object {
                    "draggingFromThisWith": null,
                    "draggingOverWith": null,
                    "isDraggingOver": false,
                  }
                }
                type="DEFAULT"
                updateViewportMaxScroll={[Function]}
              >
                <ul
                  className="map-sidebar__list"
                  data-react-beautiful-dnd-droppable="1"
                  onClick={[MockFunction]}
                >
                  <Memo(SidebarRoutesItem)
                    content="точка 1"
                    id="point-0"
                    index={0}
                    key="point-0"
                  >
                    <ConnectFunction
                      disableInteractiveElementBlocking={false}
                      draggableId="point-0"
                      index={0}
                      isDragDisabled={false}
                      shouldRespectForcePress={false}
                    >
                      <Draggable
                        disableInteractiveElementBlocking={false}
                        draggableId="point-0"
                        drop={[Function]}
                        dropAnimationFinished={[Function]}
                        index={0}
                        isDragDisabled={false}
                        lift={[Function]}
                        mapped={
                          Object {
                            "combineTargetFor": null,
                            "offset": Object {
                              "x": 0,
                              "y": 0,
                            },
                            "shouldAnimateDisplacement": true,
                            "snapshot": Object {
                              "combineTargetFor": null,
                              "combineWith": null,
                              "draggingOver": null,
                              "dropAnimation": null,
                              "isDragging": false,
                              "isDropAnimating": false,
                              "mode": null,
                            },
                            "type": "SECONDARY",
                          }
                        }
                        move={[Function]}
                        moveByWindowScroll={[Function]}
                        moveDown={[Function]}
                        moveLeft={[Function]}
                        moveRight={[Function]}
                        moveUp={[Function]}
                        shouldRespectForcePress={false}
                      >
                        <li
                          aria-roledescription="Draggable item. Press space bar to lift"
                          className="map-sidebar__list__item"
                          data-id="point-0"
                          data-index={0}
                          data-react-beautiful-dnd-drag-handle="1"
                          data-react-beautiful-dnd-draggable="1"
                          draggable={false}
                          onBlur={[Function]}
                          onDragStart={[Function]}
                          onFocus={[Function]}
                          onKeyDown={[Function]}
                          onMouseDown={[Function]}
                          onTouchStart={[Function]}
                          onTransitionEnd={null}
                          style={
                            Object {
                              "transform": null,
                              "transition": null,
                            }
                          }
                          tabIndex={0}
                        >
                          <span>
                            точка 1
                          </span>
                          <span
                            className="close-item"
                          >
                            [x]
                          </span>
                        </li>
                      </Draggable>
                    </ConnectFunction>
                  </Memo(SidebarRoutesItem)>
                  <Memo(SidebarRoutesItem)
                    content="точка 2"
                    id="point-1"
                    index={1}
                    key="point-1"
                  >
                    <ConnectFunction
                      disableInteractiveElementBlocking={false}
                      draggableId="point-1"
                      index={1}
                      isDragDisabled={false}
                      shouldRespectForcePress={false}
                    >
                      <Draggable
                        disableInteractiveElementBlocking={false}
                        draggableId="point-1"
                        drop={[Function]}
                        dropAnimationFinished={[Function]}
                        index={1}
                        isDragDisabled={false}
                        lift={[Function]}
                        mapped={
                          Object {
                            "combineTargetFor": null,
                            "offset": Object {
                              "x": 0,
                              "y": 0,
                            },
                            "shouldAnimateDisplacement": true,
                            "snapshot": Object {
                              "combineTargetFor": null,
                              "combineWith": null,
                              "draggingOver": null,
                              "dropAnimation": null,
                              "isDragging": false,
                              "isDropAnimating": false,
                              "mode": null,
                            },
                            "type": "SECONDARY",
                          }
                        }
                        move={[Function]}
                        moveByWindowScroll={[Function]}
                        moveDown={[Function]}
                        moveLeft={[Function]}
                        moveRight={[Function]}
                        moveUp={[Function]}
                        shouldRespectForcePress={false}
                      >
                        <li
                          aria-roledescription="Draggable item. Press space bar to lift"
                          className="map-sidebar__list__item"
                          data-id="point-1"
                          data-index={1}
                          data-react-beautiful-dnd-drag-handle="1"
                          data-react-beautiful-dnd-draggable="1"
                          draggable={false}
                          onBlur={[Function]}
                          onDragStart={[Function]}
                          onFocus={[Function]}
                          onKeyDown={[Function]}
                          onMouseDown={[Function]}
                          onTouchStart={[Function]}
                          onTransitionEnd={null}
                          style={
                            Object {
                              "transform": null,
                              "transition": null,
                            }
                          }
                          tabIndex={0}
                        >
                          <span>
                            точка 2
                          </span>
                          <span
                            className="close-item"
                          >
                            [x]
                          </span>
                        </li>
                      </Draggable>
                    </ConnectFunction>
                  </Memo(SidebarRoutesItem)>
                  <Memo(SidebarRoutesItem)
                    content="точка 3"
                    id="point-2"
                    index={2}
                    key="point-2"
                  >
                    <ConnectFunction
                      disableInteractiveElementBlocking={false}
                      draggableId="point-2"
                      index={2}
                      isDragDisabled={false}
                      shouldRespectForcePress={false}
                    >
                      <Draggable
                        disableInteractiveElementBlocking={false}
                        draggableId="point-2"
                        drop={[Function]}
                        dropAnimationFinished={[Function]}
                        index={2}
                        isDragDisabled={false}
                        lift={[Function]}
                        mapped={
                          Object {
                            "combineTargetFor": null,
                            "offset": Object {
                              "x": 0,
                              "y": 0,
                            },
                            "shouldAnimateDisplacement": true,
                            "snapshot": Object {
                              "combineTargetFor": null,
                              "combineWith": null,
                              "draggingOver": null,
                              "dropAnimation": null,
                              "isDragging": false,
                              "isDropAnimating": false,
                              "mode": null,
                            },
                            "type": "SECONDARY",
                          }
                        }
                        move={[Function]}
                        moveByWindowScroll={[Function]}
                        moveDown={[Function]}
                        moveLeft={[Function]}
                        moveRight={[Function]}
                        moveUp={[Function]}
                        shouldRespectForcePress={false}
                      >
                        <li
                          aria-roledescription="Draggable item. Press space bar to lift"
                          className="map-sidebar__list__item"
                          data-id="point-2"
                          data-index={2}
                          data-react-beautiful-dnd-drag-handle="1"
                          data-react-beautiful-dnd-draggable="1"
                          draggable={false}
                          onBlur={[Function]}
                          onDragStart={[Function]}
                          onFocus={[Function]}
                          onKeyDown={[Function]}
                          onMouseDown={[Function]}
                          onTouchStart={[Function]}
                          onTransitionEnd={null}
                          style={
                            Object {
                              "transform": null,
                              "transition": null,
                            }
                          }
                          tabIndex={0}
                        >
                          <span>
                            точка 3
                          </span>
                          <span
                            className="close-item"
                          >
                            [x]
                          </span>
                        </li>
                      </Draggable>
                    </ConnectFunction>
                  </Memo(SidebarRoutesItem)>
                  <Memo(SidebarRoutesItem)
                    content="точка 4"
                    id="point-3"
                    index={3}
                    key="point-3"
                  >
                    <ConnectFunction
                      disableInteractiveElementBlocking={false}
                      draggableId="point-3"
                      index={3}
                      isDragDisabled={false}
                      shouldRespectForcePress={false}
                    >
                      <Draggable
                        disableInteractiveElementBlocking={false}
                        draggableId="point-3"
                        drop={[Function]}
                        dropAnimationFinished={[Function]}
                        index={3}
                        isDragDisabled={false}
                        lift={[Function]}
                        mapped={
                          Object {
                            "combineTargetFor": null,
                            "offset": Object {
                              "x": 0,
                              "y": 0,
                            },
                            "shouldAnimateDisplacement": true,
                            "snapshot": Object {
                              "combineTargetFor": null,
                              "combineWith": null,
                              "draggingOver": null,
                              "dropAnimation": null,
                              "isDragging": false,
                              "isDropAnimating": false,
                              "mode": null,
                            },
                            "type": "SECONDARY",
                          }
                        }
                        move={[Function]}
                        moveByWindowScroll={[Function]}
                        moveDown={[Function]}
                        moveLeft={[Function]}
                        moveRight={[Function]}
                        moveUp={[Function]}
                        shouldRespectForcePress={false}
                      >
                        <li
                          aria-roledescription="Draggable item. Press space bar to lift"
                          className="map-sidebar__list__item"
                          data-id="point-3"
                          data-index={3}
                          data-react-beautiful-dnd-drag-handle="1"
                          data-react-beautiful-dnd-draggable="1"
                          draggable={false}
                          onBlur={[Function]}
                          onDragStart={[Function]}
                          onFocus={[Function]}
                          onKeyDown={[Function]}
                          onMouseDown={[Function]}
                          onTouchStart={[Function]}
                          onTransitionEnd={null}
                          style={
                            Object {
                              "transform": null,
                              "transition": null,
                            }
                          }
                          tabIndex={0}
                        >
                          <span>
                            точка 4
                          </span>
                          <span
                            className="close-item"
                          >
                            [x]
                          </span>
                        </li>
                      </Draggable>
                    </ConnectFunction>
                  </Memo(SidebarRoutesItem)>
                  <AnimateInOut
                    on={null}
                    shouldAnimate={true}
                  />
                </ul>
              </Droppable>
            </ConnectFunction>
          </Provider>
        </App>
      </ErrorBoundary>
    </DragDropContext>
  </aside>
</Memo(SidebarRoutes)>
`;
